src/main/java/org/dbos/apiary/mongo/MongoConnection.java

```diff
@@ -101,17 +102,60 @@ public boolean validate(Map<String, List<String>> writtenKeys, TransactionContex
+        ClientSession session = client.startSession();
+        TransactionBody<Boolean> txnBody = () -> {
+            for (String collectionName: writtenKeys.keySet()) {
+                if (writtenKeys.get(collectionName).size() >= 10000) {
+                    continue; // Speed up bulk-loading in benchmarks.
+                }
+                MongoCollection<Document> c = database.getCollection(collectionName);
+                for (String key: writtenKeys.get(collectionName)) {
+                    c.updateOne(session, Filters.and(
+                                    Filters.eq(MongoContext.apiaryID, key),
+                                    Filters.eq(MongoContext.beginVersion, txc.txID)
+                            ),
+                            Updates.set(MongoContext.committed, true)
+                    );
+                    c.updateMany(session, Filters.and(
+                                    Filters.eq(MongoContext.apiaryID, key),
+                                    Filters.ne(MongoContext.beginVersion, txc.txID)
+                            ),
+                            Updates.set(MongoContext.committed, false)
+                    );
+                }
+            }
+            return Boolean.TRUE;
+        };
+        session.withTransaction(txnBody, TransactionOptions.builder().writeConcern(WriteConcern.MAJORITY).readConcern(ReadConcern.SNAPSHOT).build());
+        session.close();
+    }
+
     @Override
     public void garbageCollect(Set<TransactionContext> activeTransactions) {
         long globalxmin = activeTransactions.stream().mapToLong(i -> i.xmin).min().getAsLong();
         // No need to keep track of writes that are visible to all active or future transactions.
         committedWrites.values().forEach(i -> i.values().forEach(w -> w.removeIf(txID -> txID < globalxmin)));
         // Delete old versions that are no longer visible to any active or future transaction.
-        for (String collectionName : committedWrites.keySet()) {
-            MongoCollection<Document> c = database.getCollection(collectionName);
-            c.deleteMany(
-                    Filters.lt(MongoContext.endVersion, globalxmin)
-            );
+        if (ApiaryConfig.isolationLevel == ApiaryConfig.REPEATABLE_READ) {
+            for (String collectionName : committedWrites.keySet()) {
+                MongoCollection<Document> c = database.getCollection(collectionName);
+                c.deleteMany(
+                        Filters.lt(MongoContext.endVersion, globalxmin)
+                );
+            }
+        } else {
+            assert (ApiaryConfig.isolationLevel == ApiaryConfig.READ_COMMITTED);
+            for (String collectionName : committedWrites.keySet()) {
+                MongoCollection<Document> c = database.getCollection(collectionName);
+                c.deleteMany(
+                        Filters.and(
+                                Filters.eq(MongoContext.beginVersion, globalxmin),
```

Comment:
Should this be `Filters.lt()` instead of `eq`?

Reply:
Good catch, fixed!